.travis.yml

```diff
@@ -34,5 +34,15 @@ script:
 after_success:
     - if [[ $TRAVIS_PYTHON_VERSION != pypy* ]]; then pip install coveralls --use-mirrors ; coveralls ; fi
 
+# publish new version to pypi
+deploy:
+  provider: pypi
+  skip_existing: true
+  on:
+    tags: true
```

Comment:
`on: tags: true` will publish from any tag on any branch, no?

Maybe it would be more desirable to only publish when something on `master` is tagged?  https://docs.travis-ci.com/user/deployment#conditional-releases-with-on. Also, not sure if possible, but Travis may offer a way to match the tag to only a final release (e.g. 1.2.3, not 1.2.3.dev0) as defined by PEP 440.

Reply:
> `on: tags: true` will publish from any tag on any branch, no?

Yes it will publish on any tag on any branch. But because tags are almost always published on `master`, it shouldn't cause much problems.

I think that Travis CI allows conditional deployments to be either on all tags on all branches or commits on specific branch, so publishing only on tags on `master` could be harder. It may be possible with custom `condition` but I'm now sure how exactly will this work.

> Also, not sure if possible, but Travis may offer a way to match the tag to only a final release (e.g. 1.2.3, not 1.2.3.dev0) as defined by PEP 440.

It's probably also possible with custom `condition`. But I also don't know if this is really needed because pip and PyPI will handle pre-releases and not install them unless specifically required by user.

Reply:
> I think that Travis CI allows conditional deployments to be either on all tags on all branches or commits on specific branch, so publishing only on tags on master could be harder. It may be possible with custom condition but I'm now sure how exactly will this work.

So, I am not a Travis expert at all, but reading between the lines [here](https://docs.travis-ci.com/user/deployment#conditional-releases-with-on),
it looks like this may be satisfied with:

```
  on:
    branch: staging
    tags: true
```

> It's probably also possible with custom condition. But I also don't know if this is really needed because pip and PyPI will handle pre-releases and not install them unless specifically required by user.

For this, I was more getting at the possibility that a tag is pushed that doesn't represent a PEP-440 compliant version at all, such as `git tag -a "this-is-not-a-version"` ... `git push ... --follow-tags`. I agree with you that it is fine to upload development, pre-release, and post-release tags because PyPI very much can handle them as you point out.

Reply:
I found this:

> `tags: true`: deployment is triggered if and only if `$TRAVIS_TAG` is set. Depending on your workflow, you may set `$TRAVIS_TAG` explicitly, even if this is a non-tag build when it was initiated. **This causes the branch condition to be ignored.**

However, I don't know if this means that branch condition is ignored only if you manually set `$TRAVIS_TAG` or in any case when `$TRAVIS_TAG` is set.